Stratis.Bitcoin/BlockStore/BlockRepository.cs

```diff
@@ -139,58 +141,93 @@ public Task<uint256> GetTrxBlockIdAsync(uint256 trxid)
-                        this.session.Transaction.Insert<byte[], Block>("Block", blockId.ToBytes(), block);
-
-                        if (this.TxIndex)
-                        {
-                            // index transactions
-                            foreach (var transaction in block.Transactions)
-                            {
-                                var trxId = transaction.GetHash();
-                                this.PerformanceCounter.AddRepositoryInsertCount(1);
-                                this.session.Transaction.Insert<byte[], uint256>("Transaction", trxId.ToBytes(), blockId);
-                            }
-                        }
-                    }
-                    else
-                    {
-                        this.PerformanceCounter.AddRepositoryHitCount(1);
-                    }
+            var blockDict = new Dictionary<uint256, Block>();
+
+            // Gather blocks
+            foreach (var block in blocks)
+            {
+                var blockId = block.GetHash();
+                blockDict[blockId] = block;
+            }
+
+            // Sort blocks. Be consistent in always converting our keys to byte arrays using the ToBytes method.
+            var blockList = blockDict.ToList();
+            blockList.Sort((pair1, pair2) => byteListComparer.Compare(pair1.Key.ToBytes(), pair2.Key.ToBytes()));
+
+            // Index blocks
+            foreach (KeyValuePair<uint256, Block> kv in blockList)
+            {
+                var blockId = kv.Key;
+                var block = kv.Value;
+
+                // if the block is already in store don't write it again
+                var item = this.session.Transaction.Select<byte[], Block>("Block", blockId.ToBytes());
+                if (!item.Exists)
+                {
+                    this.PerformanceCounter.AddRepositoryMissCount(1);
+                    this.PerformanceCounter.AddRepositoryInsertCount(1);
+                    this.session.Transaction.Insert<byte[], Block>("Block", blockId.ToBytes(), block);
+
+                    if (this.TxIndex)
+                        foreach (var transaction in block.Transactions)
+                            transactions.Add((transaction, block));
+                }
+                else
+                {
+                    this.PerformanceCounter.AddRepositoryHitCount(1);
                 }
+            }
 
-				this.SaveBlockHash(nextBlockHash);
-				this.session.Transaction.Commit();
-			});
-		}
+            if (this.TxIndex)
+                OnInsertTransactions(transactions);
+        }
+
+        virtual protected void OnInsertTransactions(List<(Transaction, Block)> transactions)
+        {        
+            var byteListComparer = new ByteListComparer();
 
-		private bool? LoadTxIndex()
+            transactions.Sort((pair1, pair2) => byteListComparer.Compare(pair1.Item1.GetHash().ToBytes(), pair2.Item1.GetHash().ToBytes()));
```

Comment:
Something to bear in mind is the sorting of an external list (transactions).
It would be safer to do a ToList() and sort the new list or use the OrderBy() linq extension 
Not a big deal though.